clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -52,51 +53,96 @@ public boolean hasCompatibleMagic(byte magic) {
         }
     }
 
-    private MemoryRecords convert(byte magic,
-                                  TimestampType timestampType,
-                                  CompressionType compressionType,
-                                  long logAppendTime,
-                                  List<LogRecord> records) {
-
-        LogRecord firstRecord = records.iterator().next();
-        long firstOffset = firstRecord.offset();
-        ByteBuffer buffer = ByteBuffer.allocate(estimatedSizeRecords(compressionType, records));
-        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, compressionType, timestampType,
-                firstOffset, logAppendTime, 0L, (short) 0, 0);
-        for (LogRecord record : records)
-            builder.appendWithOffset(record.offset(), record.timestamp(), record.key(), record.value());
+    private MemoryRecords upconvertNoncompressedToV2AndAbove(byte magic, TimestampType upconvertTimestampType,
+                                                             List<LogEntryAndRecords> logEntryAndRecordsList) {
+        // when converting from older magic versions, we can pack the records in a message set into a single
+        // log entry. This gives us the benefit of the more efficient packing in v2 and above, and to avoid
+        // the increased overhead for singleton message sets.
+
+        List<LogRecord> logRecords = new ArrayList<>(logEntryAndRecordsList.size());
+        for (LogEntryAndRecords logEntryAndRecords : logEntryAndRecordsList)
+            logRecords.addAll(logEntryAndRecords.records);
+
+        LogEntryAndRecords firstEntryAndRecords = logEntryAndRecordsList.get(0);
+        LogEntry firstEntry = firstEntryAndRecords.entry;
+
+        final TimestampType timestampType;
+        if (firstEntry.magic() == LogEntry.MAGIC_VALUE_V0)
+            timestampType = upconvertTimestampType;
+        else
+            timestampType = firstEntry.timestampType();
+
+        long logAppendTime = timestampType == TimestampType.LOG_APPEND_TIME ? firstEntry.maxTimestamp() : LogEntry.NO_TIMESTAMP;
+
+        ByteBuffer buffer = ByteBuffer.allocate(EosLogEntry.sizeInBytes(firstEntryAndRecords.baseOffset, logRecords));
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, firstEntry.compressionType(),
+                timestampType, firstEntryAndRecords.baseOffset, logAppendTime);
+
+        for (LogRecord logRecord : logRecords)
+            builder.append(logRecord);
         return builder.build();
     }
 
+    private ByteBuffer convertLogEntry(byte magic, ByteBuffer buffer, TimestampType upconvertTimestampType,
+                                       LogEntryAndRecords logEntryAndRecords) {
+        LogEntry entry = logEntryAndRecords.entry;
+        final TimestampType timestampType;
+        if (entry.magic() == LogEntry.MAGIC_VALUE_V0)
+            timestampType = upconvertTimestampType;
+        else
+            timestampType = entry.timestampType();
```

Comment:
Correct. The upconversion timestamp type is only needed when converting from magic 0. Later versions already have their own timestamp type which should be used. 

I'm not sure I understand the suggestion. If we add more conversion methods, it will put the burden on the caller to figure out which one is needed. What is the benefit? Reducing the parameter list by one? It is also possible that the record set has entries from multiple magic versions, so how would we handle that?

In fact, the current code only uses this method for down-conversion, so another option might be to change the name to `downconvertToMagic`, and simply raise an exception if up-conversion is requested. In that case, we could remove the second argument. Later if we have an actual need for up-conversion, we could implement `upconvertToMagic`.